src/pcnt.rs

```diff
@@ -0,0 +1,543 @@
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn event_disable(&self, evt_type: PcntEventType) -> Result<(), EspError> {
+        let evt_type: esp_idf_sys::pcnt_evt_type_t = evt_type.bits();
+        unsafe { esp!(esp_idf_sys::pcnt_event_disable(self.unit, evt_type)) }
+    }
+
+    fn only_one_event_type(evt_type: PcntEventType) -> Result<esp_idf_sys::pcnt_evt_type_t, EspError> {
+        match evt_type.iter().count() {
+            1 => Ok(evt_type.bits()),
+            _ =>Err(EspError::from(esp_idf_sys::ESP_ERR_INVALID_ARG as esp_idf_sys::esp_err_t).unwrap()),
+        }
+    }
+
+    #[doc = " @brief Set PCNT event value of PCNT unit"]
+    #[doc = ""]
+    #[doc = " @param evt_type Watch point event type."]
+    #[doc = "                 All enabled events share the same interrupt (one interrupt per pulse counter unit)."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn set_event_value(&self, evt_type: PcntEventType, value: i16) -> Result<(), EspError> {
+        let evt_type = Self::only_one_event_type(evt_type)?;
+        unsafe {
+            esp!(esp_idf_sys::pcnt_set_event_value(
+                self.unit, evt_type, value
+            ))
+        }
+    }
+
+    #[doc = " @brief Get PCNT event value of PCNT unit"]
+    #[doc = ""]
+    #[doc = " @param evt_type Watch point event type."]
+    #[doc = "                 All enabled events share the same interrupt (one interrupt per pulse counter unit)."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - i16"]
+    #[doc = "     - EspError"]
+    pub fn get_event_value(&self, evt_type: PcntEventType) -> Result<i16, EspError> {
+        let evt_type = Self::only_one_event_type(evt_type)?;
+        let mut value = 0i16;
+        unsafe {
+            esp!(esp_idf_sys::pcnt_get_event_value(
+                self.unit,
+                evt_type,
+                &mut value as *mut i16
+            ))?;
+        }
+        Ok(value)
+    }
+
+    #[doc = " @brief Get PCNT event status of PCNT unit"]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - i32"]
+    #[doc = "     - EspError"]
+    // TODO: status is a bit field!
+    pub fn get_event_status(&self) -> Result<u32, EspError> {
+        let mut value = 0u32;
+        unsafe {
+            esp!(esp_idf_sys::pcnt_get_event_status(
+                self.unit,
+                &mut value as *mut u32
+            ))?;
+        }
+        Ok(value)
+    }
+
+    #[doc = " @brief Configure PCNT pulse signal input pin and control input pin"]
+    #[doc = ""]
+    #[doc = " @param channel PcntChannel"]
+    #[doc = " @param pulse_io Pulse signal input pin"]
+    #[doc = " @param ctrl_io Control signal input pin"]
+    #[doc = ""]
+    #[doc = " @note  Set the signal input to PCNT_PIN_NOT_USED if unused."]
+    #[doc = ""]
+    #[doc = " @return"]
+    #[doc = "     - ()"]
+    #[doc = "     - EspError"]
+    pub fn set_pin(
```

Comment:
Unless there are good reasons not to, all of these pins are supposed to be passed at construction time. And then all other stuff discussed applies here as well - as in don't force the user to pass these by reference or downgrade them (by the way, passing them by `&` vs `&mut` is wrong even at the moment - as it implies that you can re-use the same pin at multiple places).

Reply:
You can us the pin in multiple places.  The encoder example for instance uses the same pins in both channels (just swapped).

Reply:
Plus based on the v5 documentation you can then "unit test" code using the  PCNT by setting the pins high/low....